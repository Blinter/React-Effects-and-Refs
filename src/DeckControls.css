/* DeckControls.css */
.deck-controls {
    text-align: center;
    margin-bottom: 20px;
  }
  
  .deck-controls button {
    padding: 10px 15px;
    margin: 0 5px;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s;
  }
  
  .deck-controls button:hover {
    background-color: #e6e6e6;
  }
  
  .remaining-cards {
    margin-top: 10px;
    font-weight: bold;
  }
  
  .auto-draw-button {
    background-color: #28a745;
    color: white;
  }
  
  .auto-draw-button:hover {
    background-color: #218838;
  }
  
  .auto-draw-button:disabled {
    opacity: 0.7;
    cursor: not-allowed;
  }