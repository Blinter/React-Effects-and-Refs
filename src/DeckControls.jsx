// DeckControls.jsx
import React from 'react';
import PropTypes from 'prop-types';
import './DeckControls.css';

function DeckControls({
    deckId,
    deckState,
    shuffle,
    clear,
    add,
    show,
    remaining,
    toggleAutoDraw,
    autoDrawEnabled
}) {
    const newCard = () =>
        deckId !== null ?
            add() :
            show();

    const RemainingCards = () => {
        return (
            <>
                <p>Remaining: {remaining}</p>
            </>
        );
    }

    const findNextAction = () => {
        if (deckState.length === 0 && deckId == null) {
            return (
                <button
                    onClick={clear}
                    aria-label='Clear the deck'
                >
                    Get New Deck
                </button>
            );
        }
        const lastCard = deckState.at(-1);
        if ((deckId != null &&
            lastCard != null &&
            lastCard.remaining === 0) ||
            (deckId == null &&
                remaining === 0)) {
            return (
                <>
                    <button
                        onClick={shuffle}
                        onMouseOver={e =>
                            e.currentTarget.style.backgroundColor = 'lightblue'}
                        onMouseOut={e =>
                            e.currentTarget.style.backgroundColor = ''}
                        aria-label="Shuffle the deck"
                    >Shuffle</button>
                    |
                    <button
                        onClick={clear}
                        onMouseOver={e =>
                            e.currentTarget.style.backgroundColor = 'lightblue'}
                        onMouseOut={e =>
                            e.currentTarget.style.backgroundColor = ''}
                        aria-label='Clear the deck'
                    >Clear</button>
                </>
            );
        }

        return (
            <>
                <button
                    onClick={newCard}
                    onMouseOver={e =>
                        e.currentTarget.style.backgroundColor = 'lightblue'}
                    onMouseOut={e =>
                        e.currentTarget.style.backgroundColor = ''}
                    aria-label="Draw a card"
                >Draw</button>
                |
                <button
                    onClick={clear}
                    onMouseOver={e =>
                        e.currentTarget.style.backgroundColor = 'lightblue'}
                    onMouseOut={e =>
                        e.currentTarget.style.backgroundColor = ''}
                    aria-label='Clear the deck'
                >Clear</button>
                |
                <button
                    onClick={toggleAutoDraw}
                    style={{
                        backgroundColor: autoDrawEnabled ?
                            'lightgreen' :
                            ''
                    }}
                    aria-label={autoDrawEnabled ?
                        'Stop Auto Draw' :
                        'Start Auto Draw'}
                >
                    {autoDrawEnabled ?
                        'Stop Auto Draw' :
                        'Start Auto Draw'}
                </button >
            </>
        );
    }

    return (
        <div className="deck-controls">
            {findNextAction()}
            <p className="remaining-cards">Remaining: {remaining}</p>
        </div>
    );
}

DeckControls.propTypes = {
    deckId: PropTypes.string,
    deckState: PropTypes.array.isRequired,
    shuffle: PropTypes.func.isRequired,
    clear: PropTypes.func.isRequired,
    add: PropTypes.func.isRequired,
    show: PropTypes.func.isRequired,
    remaining: PropTypes.number.isRequired,
    toggleAutoDraw: PropTypes.func.isRequired,
    autoDrawEnabled: PropTypes.bool.isRequired
};

export default DeckControls;