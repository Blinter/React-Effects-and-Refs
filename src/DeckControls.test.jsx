import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import DeckControls from './DeckControls';

describe('DeckControls component', () => {
    const mockProps = {
        deckId: 'test-deck-id',
        deckState: [],
        shuffle: jest.fn(),
        clear: jest.fn(),
        add: jest.fn(),
        show: jest.fn(),
        remaining: 52,
        toggleAutoDraw: jest.fn(),
        autoDrawEnabled: false,
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('matches snapshot', () => {
        const { container } =
            render(<DeckControls {...mockProps} />);
        expect(container).toMatchSnapshot();
    });

    test('renders Auto Draw button', () => {
        render(<DeckControls {...mockProps} />);

        const autoDrawButton = screen.getByRole('button', { name: /auto draw/i });
        expect(autoDrawButton).toBeInTheDocument();
    });

    test('displays correct Auto Draw button text when disabled', () => {
        render(<DeckControls {...mockProps} />);

        const autoDrawButton = screen.getByRole('button', { name: /start auto draw/i });
        expect(autoDrawButton).toHaveTextContent('Start Auto Draw');
    });

    test('displays correct Auto Draw button text when enabled', () => {
        mockProps.autoDrawEnabled = true;
        render(<DeckControls {...mockProps} />);

        const autoDrawButton = screen.getByRole('button', { name: /stop auto draw/i });
        expect(autoDrawButton).toHaveTextContent('Stop Auto Draw');
    });

    test('calls toggleAutoDraw when Auto Draw button is clicked', () => {
        render(<DeckControls {...mockProps} />);

        const autoDrawButton = screen.getByRole('button', { name: /auto draw/i });
        fireEvent.click(autoDrawButton);

        expect(mockProps.toggleAutoDraw).toHaveBeenCalledTimes(1);
    });

    test('applies correct style when Auto Draw is enabled', () => {
        mockProps.autoDrawEnabled = true;
        render(<DeckControls {...mockProps} />);

        const autoDrawButton = screen.getByRole('button', { name: /stop auto draw/i });
        expect(autoDrawButton).toHaveStyle('background-color: lightgreen');
    });

    test('renders remaining cards count', () => {
        render(<DeckControls {...mockProps} />);

        const remainingCount = screen.getByText(/Remaining: \d+/);
        expect(remainingCount).toBeInTheDocument();
        expect(remainingCount).toHaveTextContent('Remaining: 52');
    });

    test('updates remaining cards count when prop changes', () => {
        const { rerender } = render(<DeckControls {...mockProps} />);

        rerender(<DeckControls {...mockProps} remaining={30} />);

        const remainingCount = screen.getByText(/Remaining: \d+/);
        expect(remainingCount).toHaveTextContent('Remaining: 30');
    });
});