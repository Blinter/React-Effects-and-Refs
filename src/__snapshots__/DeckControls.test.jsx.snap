// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckControls component matches snapshot 1`] = `
<div>
  <div
    class="deck-controls"
  >
    <button
      aria-label="Draw a card"
    >
      Draw
    </button>
    |
    <button
      aria-label="Clear the deck"
    >
      Clear
    </button>
    |
    <button
      aria-label="Start Auto Draw"
    >
      Start Auto Draw
    </button>
    <p
      class="remaining-cards"
    >
      Remaining: 
      52
    </p>
  </div>
</div>
`;
